### ISSUE TITLE
6.2.7 and 6.2.8 levels are unclear

### ISSUE THREAD

**Issue by Racater** [OWASP/ASVS#888]
The section 6.2 states that:
> Although this section is not easily penetration tested, developers should consider this entire section as mandatory even though L1 is missing from most of the items.

In the rest of the OWASP ASVS, non pen-testable L1 items are always marked as L2.

However, items [6.2.7](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x14-V6-Cryptography.md#user-content-v62-algorithms) and [6.2.8](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x14-V6-Cryptography.md#user-content-v62-algorithms) are L3 and not L2. Is there any reason for this? It seems contradictory.

Should these items be considered L1 (as the section text suggests) or L3 (as they are L3, more specific than just non pen-testable). This may need to be clarified.

Or is there something I missed?

**Comment by Racater** [comment]
The question is: should [6.2.7](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x14-V6-Cryptography.md#user-content-v62-algorithms) and [6.2.8](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x14-V6-Cryptography.md#user-content-v62-algorithms) be L1, L2 or L3. Any opinion?

**Comment by jmanico** [comment]
I would like to move 6.2.7 to L2 but leave 6.2.8 as L3.  Just my instinct. If you want to move both of these to L2 I support it. Let's get a PR from you, @Racater !

**Comment by Racater** [comment]
Currently, 6.2.7 and 6.2.8 are both L1 (mandatory, but non pen-testable), because of the section text. Should we remove them from L1?

**Comment by Sjord** [comment]
> Verify that encrypted data is authenticated via signatures, authenticated cipher modes, or HMAC to ensure that ciphertext is not altered by an unauthorized party.

This can sometimes be tested, if the encrypted data passes through the client. For example, an encrypted cookie value. If flipping a bit in the cookie value does not trigger an error, it is not protected. Unprotected encrypted data is somewhat common, and I have found it in a couple of pentests. I would say L2.

> Verify that all cryptographic operations are constant-time, with no 'short-circuit' operations in comparisons, calculations, or returns, to avoid leaking information.

"Constant-time" is almost impossible to test remotely. "Short-circuit" may be possible, for example in padding attacks. I would say L3.

> Although this section is not easily penetration tested, developers should consider this entire section as mandatory even though L1 is missing from most of the items.

This should be removed, both here and in 6.4. It blurs the meaning of the levels, by saying that L3 reqs should be treated as L1. If pentesters check the work of developers, they should agree of which reqs are which levels, so marking reqs as L1 for developers and L3 for pentesters doesn't work. [Chapter 3](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x03-Using-ASVS.md#application-security-verification-levels) already explains that the levels are for pentesting and that this is not a perfect categorization for developers.

**Comment by Racater** [comment]
> Although this section is not easily penetration tested, developers should consider this entire section as mandatory even though L1 is missing from most of the items.

> This should be removed, both here and in 6.4. It blurs the meaning of the levels, by saying that L3 reqs should be treated as L1.

I agree with you and I am also in favor of clearer levels. This was probably added to have L1 cover the complete OWASP top 10 (here probably #&#x2060;9: Using Components with Known Vulnerabilities).

There are many places in the standard where the levels are not the same for pentesters and developers: [2.4](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x11-V2-Authentication.md#user-content-v24-credential-storage-requirements), [2.10](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x11-V2-Authentication.md#user-content-v210-service-authentication-requirements), [5.1](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x13-V5-Validation-Sanitization-Encoding.md#user-content-v51-input-validation-requirements), [6.2](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x14-V6-Cryptography.md#user-content-v62-algorithms), [6.4](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x14-V6-Cryptography.md#user-content-v64-secret-management), [7.1](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x15-V7-Error-Logging.md#user-content-v71-log-content-requirements), [7.2](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x15-V7-Error-Logging.md#user-content-v72-log-processing-requirements) and [7.3.1](https://github.com/OWASP/ASVS/blob/master/4.0/en/0x15-V7-Error-Logging.md#user-content-v73-log-protection-requirements). What are your opinion on this? It is probably better to have a global policy towards this use case.

**Comment by jmanico** [comment]
I'm not sure what is best but if between @Racater and @Sjord you can figure out a level adjustment, I am watching for PR's every day and will merge fast. When you are both ready please seize the moment and PR us! :)

Level 1 - mostly testable and easy on developers
Level 2 - 75% testable and modest to developers
Level 3 - 50% or so testable and challenging to developers

When something does not easily fit - then please use your best judgement.

**Comment by jmanico** [comment]
> Currently, 6.2.7 and 6.2.8 are both L1 (mandatory, but non pen-testable), because of the section text. Should we remove them from L1?

These are both L3 not L1 right now.
